ceph-fuse: link to libfuse3 and pass `-o big_writes` to libfuse if libfuse < 3.0.0

lxbsz opened the pull request:
The big_writes is deprecated since libfuse 3.0. But if ceph-fuse is
linked to libfuse < 3.0, without big_writes option enabled, the write
performance is very poor.

Fixes: https://tracker.ceph.com/issues/44885
Signed-off-by: Xiubo Li <xiubli@redhat.com>


<!--
Thank you for opening a pull request!  Here are some tips on creating
a well formatted contribution.

Please give your pull request a title like "[component]: [short description]"

This is the format for commit messages:

"""
[component]: [short description]

[A longer multiline description]

Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
Signed-off-by: [Your Name] <[your email]>
"""

The Signed-off-by line is important, and it is your certification that
your contributions satisfy the Developers Certificate or Origin.  For
more detail, see SubmittingPatches.rst.

The component is the short name of a major daemon or subsystem,
something like "mon", "osd", "mds", "rbd, "rgw", etc. For ceph-mgr modules,
give the component as "mgr/<module name>" rather than a path into pybind.

For more examples, simply use "git log" and look at some historical commits.

This was just a quick overview.  More information for contributors is available here:
https://raw.githubusercontent.com/ceph/ceph/master/SubmittingPatches.rst

-->
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

ukernel commented on src/common/options.cc:
```diff
@@ -8340,8 +8340,13 @@ std::vector<Option> get_mds_client_options() {
+#else
+    .set_default(true)
+    .set_description("enable big_writes in libfuse < 3.0.0"),
+#endif
```
how about set this option to true by default, make ceph-fuse ignore this option if linked to libfuse3

lxbsz replied:
Yeah, sounds better.

lxbsz replied:
@ukernel Fixed it. Thanks.

gregsfortytwo commented:
Please remember to add the right labels to your PRs so they filter correctly! :)

lxbsz commented:
> Please remember to add the right labels to your PRs so they filter correctly! :)

Okay, before It seemed I had no permission to set it, but it works now. Will do that for the new ones in future.

lxbsz commented:
@ukernel  @tchaikov 
This has confilict with supporting to the libfuse3 one, I added the linking to libfuse3 here.

@gregsfortytwo
I removed the wip-grep-testing.
Thanks

ukernel commented on src/os/FuseStore.cc:
```diff
@@ -420,11 +434,19 @@ static int os_readdir(const char *path,
+      filler(buf, "type", NULL, 0, flags);
+#else
       filler(buf, "type", NULL, 0);
+#endif
```
it's better to define a helper function for filer.

lxbsz replied:
Sure, will add one.

ukernel commented on src/os/FuseStore.cc:
```diff
@@ -1127,7 +1195,9 @@ static int os_statfs(const char *path, struct statvfs *stbuf)
   readlink: 0,
+#if FUSE_VERSION < FUSE_MAKE_VERSION(3, 0)
   getdir: 0,
+#endif
```
filled with 0 by default.   this is no needed

lxbsz replied:
Will fix it.

lxbsz replied:
@ukernel 
 
I am afraid we couldn't delete it due to strict check in makefile, please see

https://github.com/ceph/ceph/pull/34531#issuecomment-615821206

ukernel commented on src/os/FuseStore.cc:
```diff
@@ -1138,7 +1208,9 @@ static struct fuse_operations fs_oper = {
   truncate: os_truncate,
+#if FUSE_VERSION < FUSE_MAKE_VERSION(3, 0)
   utime: 0,
+#endif
```
same here

tchaikov commented on cmake/modules/Findfuse3.cmake:
```diff
@@ -0,0 +1,34 @@
+# This module can find FUSE Library
```
instead of replicating `Findfuse.cmake`, i think a better approach is to add fuse3 as another suffix when finding FUSE library. could you help test and review #34604 ?

lxbsz replied:
sure, will test it.

ukernel commented:
also see https://github.com/ceph/ceph/pull/34604

lxbsz commented:
> also see #34604

@ukernel The #34604 will only handle the Cmake stuff, and we still need this one.

ukernel commented:
Sorry I accidentally closed this PR.   I think tchaikov's cmake change is better than this one's

lxbsz commented:
> Sorry I accidentally closed this PR. I think tchaikov's cmake change is better than this one's

Yeah, agree. And now I am testing it together with this one. And checked that code, looks good to me.

lxbsz commented:
This PR will depend on #34604.

tchaikov commented:
#34604 has been merged.

gregsfortytwo commented:
Well the earlier version of this passed testing, will pull it in to the next one again. :)

gregsfortytwo commented:
@lxbsz I assume the force-push was just to rebase, no code changes that would invalidate the existing reviews?

gregsfortytwo commented:
@lxbsz while it passed the Jenkins "make check", pulling it into an integration branch resulted in build failures on Bionic
https://shaman.ceph.com/builds/ceph/wip-greg-testing-417-2/7e58a389efd0c8f4a8ce5e2661acbca8f2f712db/default/202328/

> /build/ceph-16.0.0-896-g7e58a38/src/os/FuseStore.cc:1179:1: sorry, unimplemented: non-trivial designated initializers not supported

Skimming over things, I assume it's because os_truncate now takes a more complicated type in libfuse3.

lxbsz commented:
> @lxbsz I assume the force-push was just to rebase, no code changes that would invalidate the existing reviews?

Yeah, no confilict will happen with @tchaikov 's PR.

lxbsz commented:
> @lxbsz while it passed the Jenkins "make check", pulling it into an integration branch resulted in build failures on Bionic
> https://shaman.ceph.com/builds/ceph/wip-greg-testing-417-2/7e58a389efd0c8f4a8ce5e2661acbca8f2f712db/default/202328/
>
This should be the uuid issue on Bionic, and we can see the same issue for others, like:

https://shaman.ceph.com/builds/ceph/jaeger-rebase-8.0/98688e77de392377931a3dc8f04b413dd484a69c/default/202107/

 
> > /build/ceph-16.0.0-896-g7e58a38/src/os/FuseStore.cc:1179:1: sorry, unimplemented: non-trivial designated initializers not supported
> 
> Skimming over things, I assume it's because os_truncate now takes a more complicated type in libfuse3.

Yeah, in libfuse3 the options has changed much and are split into 2 groups, one of the conn and one for mounting as usual, I will spend more time on the libfuse3 and check them in detail, this PR will do the basic support the libfuse3.

lxbsz commented:
> > @lxbsz while it passed the Jenkins "make check", pulling it into an integration branch resulted in build failures on Bionic
> > https://shaman.ceph.com/builds/ceph/wip-greg-testing-417-2/7e58a389efd0c8f4a8ce5e2661acbca8f2f712db/default/202328/
> > > /build/ceph-16.0.0-896-g7e58a38/src/os/FuseStore.cc:1179:1: sorry, unimplemented: non-trivial designated initializers not supported
> > 
> > 
> > Skimming over things, I assume it's because os_truncate now takes a more complicated type in libfuse3.
> 
> Yeah, in libfuse3 the options has changed much and are split into 2 groups, one of the conn and one for mounting as usual, I will spend more time on the libfuse3 and check them in detail, this PR will do the basic support the libfuse3.

This dues to some members are removed in linfuse3, and I jsut delete them if not defined, will fix it.

```
/build/ceph-16.0.0-874-g213214f/src/os/FuseStore.cc:1179:1: sorry, unimplemented: non-trivial designated initializers not supported
 };
 ^

/build/ceph-16.0.0-874-g213214f/src/os/FuseStore.cc:1179:1: sorry, unimplemented: non-trivial designated initializers not supported
/build/ceph-16.0.0-874-g213214f/src/os/FuseStore.cc:1179:1: sorry, unimplemented: non-trivial designated initializers not supported

```

lxbsz commented:
@gregsfortytwo It's okay now.

https://shaman.ceph.com/builds/ceph/wip-lxb-libfuse3-2020-04-18-2/c96019781b7a94b2e18c4cae841fb1205f7df106/default/202451/

ukernel commented:
LGTM

lxbsz commented:
The integration test:
http://pulpito.ceph.com/xiubli-2020-04-21_03:10:12-fs:basic_functional-wip-lxb-libfuse3-2020-04-18-2-distro-basic-smithi/

gregsfortytwo commented:
http://pulpito.front.sepia.ceph.com/gregf-2020-04-21_21:12:13-fs-wip-greg-testing-421-1-distro-basic-smithi/ looks good; few infra bugs and known issues
